core/benches/scheduler.rs

```diff
@@ -0,0 +1,494 @@
+}
+
+struct SimpleTransferTxBuilder;
+impl TestTxsBuilder for SimpleTransferTxBuilder {
+    fn prepare_instructions(&self, payer: &Keypair) -> Vec<Instruction> {
+        let to_pubkey = Pubkey::new_unique();
+        vec![system_instruction::transfer(&payer.pubkey(), &to_pubkey, 1)]
+    }
+}
+
+struct ManyTransfersTxBuilder;
+impl TestTxsBuilder for ManyTransfersTxBuilder {
+    fn prepare_instructions(&self, payer: &Keypair) -> Vec<Instruction> {
+        const MAX_TRANSFERS_PER_TX: usize = 35;
+        let to = Vec::from_iter(
+            std::iter::repeat_with(|| (Pubkey::new_unique(), 1)).take(MAX_TRANSFERS_PER_TX),
+        );
+        system_instruction::transfer_many(&payer.pubkey(), &to)
+    }
+}
+
+struct PackedNoopsTxBuilder;
+impl TestTxsBuilder for PackedNoopsTxBuilder {
+    fn prepare_instructions(&self, _payer: &Keypair) -> Vec<Instruction> {
+        // Creating noop instructions to maximize the number of instructions per
+        // transaction. We can fit up to 355 noops.
+        const MAX_INSTRUCTIONS_PER_TRANSACTION: usize = 355;
+        let program_id = Pubkey::new_unique();
+        (0..MAX_INSTRUCTIONS_PER_TRANSACTION)
+            .map(|_| Instruction::new_with_bytes(program_id, &[], vec![]))
+            .collect()
+    }
+}
+
+struct BenchContainer<Tx: TransactionWithMeta> {
+    container: TransactionStateContainer<Tx>,
+}
+
+impl<Tx: TransactionWithMeta> BenchContainer<Tx> {
+    fn new(capacity: usize) -> Self {
+        Self {
+            container: TransactionStateContainer::with_capacity(capacity),
+        }
+    }
+
+    fn fill_container(&mut self, transactions: impl Iterator<Item = Tx>) {
+        for transaction in transactions {
+            let compute_unit_price = transaction
+                .compute_budget_instruction_details()
+                .sanitize_and_convert_to_compute_budget_limits(
+                    &agave_feature_set::FeatureSet::default(),
+                )
+                .unwrap()
+                .compute_unit_price;
+
+            // NOTE - setting transaction cost to be `0` for now, so it doesn't bother block_limits
+            // when scheduling.
+            const TEST_TRANSACTION_COST: u64 = 0;
+            if self.container.insert_new_transaction(
+                transaction,
+                MaxAge::MAX,
+                compute_unit_price,
+                TEST_TRANSACTION_COST,
+            ) {
+                unreachable!("test is setup to fill the Container to fullness");
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct BenchStats {
+    bench_iter_count: usize,
+    num_of_scheduling: usize,
+    // worker reports:
+    num_works: Arc<AtomicUsize>,
+    num_transaction: Arc<AtomicUsize>, // = bench_iter_count * container_capacity
+    tracer_placement: Arc<AtomicUsize>, // > 0
+    // from scheduler().result:
+    num_scheduled: usize, // = num_transaction
+}
+
+impl BenchStats {
+    fn print_and_reset(&mut self) {
+        println!(
```

Comment:
wonder if we can gate this printing so that it only prints if i request it somehow (`log::debug!`?). These values shouldn't really change significantly between runs of the benchmark - when using the benchmark we'll primarily be interested in the differences in performance since primarily these benches are not used for **logical** changes but to help driver performance improvements.

Reply:
Agree that bench essentially is for performance, less for behavior. On the other hand, the iterative nature of benching provides a good opportunity to observe scheduler behaviors. Something I was curious to study, could also be insightful than just blackbox.

I can suggest two options:
1. remove `benchState` entirely from this PR, keep is in my private repo, manually apply it when I want to see more. This is the clean approach, and keeps benches in consistent standard.
2. if people also found the Stats can be useful, we can use env variable like `ENABLE_SCHEDULER_STATS=true cargo bench ...`, then add flag in code to enable/disable states collecting and printing. This is bit unusual for bench project.

I'll be happy for both, wdyt?

Reply:
either is fine to me - would prefer some common argument such as debug logging instead of a new env-var though if we do want to go that path

Reply:
decided to remove those from this PR @ a64dd8a17e55da1314f2516a6501676e76a91da0